### ISSUE TITLE
Create annotations using only keyboard

### ISSUE THREAD

**Issue by klemay** [hypothesis/product-backlog#447]
### Problem you are trying to address with this feature
We need to enable users who don't/can't use a mouse to use our product. This is an accessibility issue for users with disabilities, and a productivity issue for other users who prefer to work from their keyboard as much as possible. 

### Your solution
From user-created issue: https://github.com/hypothesis/product-backlog/issues/446
"In general, provide option to be fully keyboard driven - remove requirement to use mouse. Allow to work with current text selection regardless of how established (ie shift arrow keys extend selection) - provide keyboard shortcuts for everything (i.e. meta-a: edit-annotation-at-selection; meta-h: toggle-highlight-selection)."

Adding this from accessibility guidelines: 
Some functionality is not accessible with a keyboard. Elements that are not accessible with a
keyboard are also not operable with screen readers. (e.g., Annotate and Highlight buttons in the adder) - related to 2.4.7 (A)

Requirement: 
- 2.4.7: Focus Visible
- [ ] In some cases keyboard focus is not visible. (A)

**Comment by klemay** [comment]
Today I learned that Chrome allows for setting custom keyboard shortcuts for extensions, and developers are able to include some defaults:

<img width="702" alt="screen shot 2018-09-03 at 12 03 27 pm" src="https://user-images.githubusercontent.com/12087705/44995545-6f673480-af71-11e8-9878-d8e77698460d.png">

I advised [a user in Zendesk](https://hypothesis.zendesk.com/agent/tickets/2968) on how to set a shortcut for enabling/disabling the Chrome extension, but we might want to explore including some defaults in the future.

**Comment by ajpeddakotla** [comment]
| User Engagement | Building Community | Level of Effort |
|---|---|---|
| 3 | 3 | 1 |

**Comment by maciejmotyka** [comment]
There are also users that can use mouse but prefer not to.

**Comment by xolotl** [comment]
For reference: adding [a link to the presentation that @jccr gave at I Annotate 2019](https://docs.google.com/presentation/d/1uR6lDou8eKsH5eXqJlmf3SGPvqdDb0MUA3Xry-2VFak/edit?usp=sharing) about using the keyboard to select/anchor annotations.

**Comment by andrewsanchez** [comment]
My method for getting keyboard only annotations requires using the Vimium extension for Chrome and specifically Vimium's [visual mode, caret mode](https://github.com/philc/vimium/wiki/Visual-Mode), and [find mode](https://github.com/philc/vimium/wiki/Find-Mode).  It seems to work really well so far.  You will need to install Vimium and familiarize yourself with the extension as well as the basics of vim keyboard navigation before this will make sense.  It might sound like a lot, but it really pays off to learn this, imo.   I hope this helps somebody!

**Initial text selection**
First, you need to place your cursor  at the beginning of your desired text selection.  There are a couple of ways to do it. 

- Method 1
  - Enter find mode by pressing `/`
  - Search for the text you'd like select.
    - I.e., search for the first `n` characters at the beginning of the text you'd like to select.

- Method 2
  - Enter visual mode by pressing `v`
    - Optional: switch to caret mode by pressing `c`.  This lets you move the cursor with vim-style navigation without selecting text. 

 I prefer method 2 because it is quicker and less error prone.

**Select the rest of the text**
Once your cursor is in the correct place, you need to select the rest of the text to complete your highlight or annotation. 

- If you are in caret mode, press `v` to go back to visual mode.
- Use vim-style shortcuts to select text.  These work in caret mode as well for moving the cursor.

For example:
  - `)` to move forward by a complete sentence, `(` to move back
  - `}` to move forward by a complete paragraph, `{` to move back

- Exit visual mode by pressing `esc`
- Press `h` to highlight or `a` to annotate using hypothesis

I've noticed inconsistent behavior with hypothesis registering the `h` or `a` shortcuts after hitting `esc` to exit out of visual mode.  Hitting the hypothesis shortcuts `h` or `a` immediately after hitting `esc` seems to solve the problem most of the time.  Also, sometimes you will encounter a conflict between Vimium and hypothesis.  It seems visual mode needs to be deactivated by pressing `esc` so that hypothesis will register the `h` an `a` shortcuts.  

Notes:
 - I've only tested this on web pages in Chrome (not pdfs). 
- EDIT:  You must unmap `h` in Vimium for this to work.  `h` will still work while in visual or caret mode to move the cursor to the left.

**Comment by klemay** [comment]
@andrewsanchez awesome, thanks for sharing! 

Those `h` and `a` shortcuts were shipped pretty recently, and we still have some documentation and expansion work ahead of us. It's great to see that you've been able to use them. FWIW, I believe there's an experimental capability in Firefox and Chrome that allows for text selection without the use of an extension - we plan to explore and document that workflow.

**Comment by robertknight** [comment]
I think we can close this issue now. Assuming the user has their browser/OS set up to enable them to make a text selection in an HTML page or PDF using only the keyboard, they can create annotations and highlights using only the keyboard:

- Select text to make the "Annotate"/"Highlight" buttons appear and enable the shortcuts for creating annotations
- To create an annotation or highlight press "a" or "h"
- Use the tab keys or screen reader navigation keys to move focus around controls in the sidebar. We now show a focus ring around all controls in the sidebar when using keyboard navigation (via CSS `focus-visible`)

**Comment by andrewsanchez** [comment]
@klemay I'd be happy to help with evaluating and documenting approaches to keyboard only usage.  Feel free to ping me to discuss.

**Comment by klemay** [comment]
@andrewsanchez to follow up on your generous offer: as of now we only come in after the text selection is made, by allowing users to press `a` or `h` on their keyboard to make an annotation or highlight, respectively.

Keyboard-only text selection is built into many browsers & most screen readers (it's needed for things like copy and paste) so at the moment we're not providing affordances outside of that. We might develop additional controls in the future. 

I've created a KB article: [Creating annotations using only your keyboard](https://web.hypothes.is/help/creating-annotations-using-only-your-keyboard/). I link out to instructions for a few major browsers, and advise users to follow directions for text selection for any other browser or AT tool. 

I [annotated the KB article](https://hyp.is/pnn0LHkOEeqDMjeokYocXA/web.hypothes.is/help/creating-annotations-using-only-your-keyboard/) linking to your comment on this issue. We'd welcome you replying there if you feel there's more info to share that isn't in your Github comment. Either way, thanks again for your offer to help us make Hypothesis more usable for everyone!

**Comment by cjnchk** [comment]
This shortcuts conflict with vimium, It would be perfect if can customize.

**Comment by maciejmotyka** [comment]
> This shortcuts conflict with vimium, It would be perfect if can customize.

@cjnchk You can use vimium's `passNextKey` command to tell vimium that the next key you press should not be intercepted and should go to the website (or hypothes.is). Alternatively you can use vimium's `insert mode` to disable vimium until the `<Esc>` is pressed. The only difference here is that `passNextKey` will pass only one key to the website, whereas `insert mode` will pass all the keys pressed until you disable it with `<Esc>`.

I have mapped `insert mode` to `i` and `passNextKey` to `a` (you can map to different keys obviously).
First, I use vimium's `visual mode` to highlight text and press `<Enter>`. Now I have 2 options:
1. press `a a` to annotate or `a h` to highlight (using `passNextKey`)
2. press `i a <Esc>` to annotate or `i h <Esc>` to highlight (using `insert mode`)

You can find out how to map keys to commands and more about `passNextKey` and `insert mode` in [vimium's wiki](https://github.com/philc/vimium/wiki/).

**Comment by cjnchk** [comment]
@maciejmotyka I use the visual mode to highlight,  but the highlight disappeared after I pressed \<Enter\>,  instead it performed a copy operation, which prevents me from performing subsequent operations. 

Is there any other way to exit the visual mode without making the highlight disappear?

My inner thought is that these methods are too circumstantial, not good, changing the shortcut keys is the right way.😆

update:
I understand that for the problem I said, when selecting text through visual mode, it must start in search mode, not in caret mode.😓

**Comment by boehs** [comment]
One issue, my browser seems to be blocking vimium from running on extension pages, has anyone run into this? Perhaps native bindings are proper?
